# Utilizing-statistical-methods-to-discover-genetic-variants-in-Brain-Disorders

Most current association studies focus on a single biological unit. In our work, we developed an Image Expression
Directed Linear Mixed Model (IEDLMM) which utilizes informative weights learned from training
genetically predictive models for brain images using a linear mixed model and for gene expressions
using a Bayesian Sparse Linear Mixed Model, to estimate the polygenic term in a linear mixed
model. Through Simulations we have proven that, IEDLMM exhibits higher power than current
methods while keeping the type-I error rates under control. By leveraging the UK Biobank imagederived
phenotypes (IDPs) and GTEx gene expression data, the IEDLMM identified 15 unique genes
related to brain disorders across four datasets which are validated through DisGeNET functional
annotations proving the efficacy of IEDLMM compared to existing methods. The creation of
IEDLMM paves the way for additional exploration in the integration of multiple omics data within
a single framework. This method not only improves the credibitility of the results but also furthers
our knowledge in the field, laying a foundation for future research efforts.

<img width="543" height="615" alt="Screenshot 2025-08-07 at 1 59 16 PM" src="https://github.com/user-attachments/assets/2b2c6f4b-7649-46ac-b5a0-2b639082df3a" />

Steps to carry out:
1) Data for real data analysis was sourced from multiple data repositoaries, which include GTEX
genotype data, UKBiobank data, Dbgap, and MSSNG. A pre-trained model on expression-genotype
data utilized gene expression data generated by RNA-seq obtained from the Genotype-Tissue Expression
Project version 8 (GTEx v8). Pre trained BSLMM was used with the default parameters to extract the weights. The genome coordinates for this model were based on the GRCh38 format.
2) An image-genotype pre-trained model was run on UK BioBank data, which comprises 33,553 individuals with 885 IDPs.weight extraction is performed from this model, to carry out weight integration with expression data, genotype coordinates are lifted over to GRCh38 format using the liftover tool.
3) The kinshipwe function in javamix5 was utilized to construct the weighted kinship matrix.
4) Subsequently, association tests were conducted using the emmax function in javamix5.


References:

Quan Long et al. JAWAMix5: an out-of-core HDF5-based java implementation of wholegenome
association studies using mixed models. Bioinformatics, 29(9):1220–1222, May 2013.

Jingni He et al. A statistical method for image-mediated association studies discovers genes
and pathways associated with four brain disorders. The American Journal of Human Genetics,
111(1):48–69, January 2024.

Qing Li et al. An expression-directed linear mixed model discovering low-effect genetic variants.
GENETICS, 226(4):iyae018, April 2024.
